Inclui aplicação para gerir PID v2 e v3 no contexto do Upload

robertatakenaka opened the pull request:
#### O que esse PR faz?
Inclui aplicação para gerir PID v2 e v3 no contexto do Upload. Esta é uma versão local. Possivelmente teremos o pid manager local (no upload) e o pid manager central (no core)

#### Onde a revisão poderia começar?
Por commits

#### Como este poderia ser testado manualmente?
Deve-se ter um registro de FilesStorageConfiguration ~, cujo `name='pid_provider'`.~
Para isso, faça o seguinte:

Como houve várias mudanças nos modelos, talvez pode ser necessário apagar as bases relacionadas com collection e migration.
Execute os comandos:
```console
make django_makemigrations
make django_migrate
make up
```
Renomeie o arquivo .bigbang.example.for.dev.purpose.only para .bigbang e adapte os valores das variáveis. Isso é para agilizar os testes em desenvolvimento. Esta etapa é um atalho. O fluxo normal requer preencher os formulários de Coleção, Novo site, Site Clássico, Minio e configuração de migração.

Entre no sistema.
Acesse o menu Tasks > Periodic tasks
Crie uma task nova, preenchendo com nome (a sua escolha), selecione a task "start", preencha o campo CRONTAB SCHEDULE, com o valor default. Caso não esteja habilitado, crie um no menu Tasks > Crontab, preenchendo hora e minuto, com qualquer valor.
Na listagem de tarefas, clique no botão RUN da tarefa "start".
Isso fará com que o arquivo ".bigbang" seja lido, os registros de Coleção, **Novo site**, Site Clássico, Minio (**FilesStorageConfiguration**) e configuração de migração sejam preenchidos, as tarefas de migração de dados de periódicos e fascículos sejam agendadas.

Do módulo pid_provider/controller.py, teste as 3 funções:
```python
def get_registered_xml(xmltree):
def request_document_ids_for_xml_uri(xml_uri, user_id):
def request_document_ids_for_zip(xml_zip_file_path, user_id):
```

#### Algum cenário de contexto que queira dar?
~WIP. Considera que todos os artigos já estão no upload.~

Falta a "cola" entre migração e pid provider, entre ingresso e pid provider. ~, falta considerar que nem todos os artigos publicados estão na base de dados do namespace pid_provider.~
Na ausência do artigo no sistema Upload, há uma sincronização com o banco do site novo (configurado no bigbang). Tenta trazer os artigos de lá e ingressar no pid_provider.

### Screenshots
n/a

#### Quais são tickets relevantes?
n/a

### Referências
n/a

robertatakenaka committed:
Cria app pid_provider

robertatakenaka committed:
Cria módulo para lidar com questões de XML SPS, arquivos XML (compactados ou não), etc

robertatakenaka committed:
Adiciona módulo de exceções

robertatakenaka committed:
Adiciona modelos para representar partes do XML e o link para o XML

robertatakenaka committed:
Acrescenta name em FilesStorageConfiguration e cria funções para obter, criar, atualizar este modelo

robertatakenaka committed:
Cria funções para receber XML de vários modos: zip ou uri e solicitar ou validar PIDs.
Também cria funções para verificar se XML já está registrado no PID provider.

robertatakenaka committed:
Adiciona módulo que gera o pid v3

robertatakenaka committed:
Faz correções

robertatakenaka committed:
Adiciona pid_provider como app

robertatakenaka committed:
Cria função para sincronizar dados de registros do site novo para o sistema do pid provider

robertatakenaka committed:
Faz com que a migração consulte o pid provider

robertatakenaka committed:
Deixa em libs as funções mais genéricas referentes ao XML SPS e em pid_provider aquelas do contexto do pid provider

robertatakenaka committed:
Adiciona o atributo v3 para XMLFile

robertatakenaka committed:
Adiciona o atributo latest para indicar que é a versão mais recente do XML

robertatakenaka committed:
Inclui o link para o xml no retorno da consulta de um documento no site novo

robertatakenaka committed:
Adiciona exceções

rafaelpezzuto commented on pid_provider/xml_sps_utils.py:
```diff
@@ -0,0 +1,438 @@
+                     (xml_sps_file_path, xml_file_path, content[:100]))
+        zf.writestr(xml_file_path, content)
+
+    return xml_sps_file_path
```
@robertatakenaka Já que o destino do conteúdo é o próprio parâmetro `xml_sps_file_path`, não seria mais útil retornar se o arquivo gravado existe ou se é um arquivo, como você fez no método seguinte?

`return os.path.exists(xml_sps_file_path)` ou `return os.path.isfile(xml_sps_file_path)`

rafaelpezzuto commented on pid_provider/xml_sps_utils.py:
```diff
@@ -0,0 +1,438 @@
+
+    @property
+    def v2_prefix(self):
+        return f"S{self.journal_issn_electronic or self.journal_issn_print}{self.pub_year}"
```
@robertatakenaka Dúvida: a prioridade na formação do v2_prefix é sempre do ISSN electronic?

rafaelpezzuto commented on pid_provider/models.py:
```diff
@@ -0,0 +1,161 @@
+        return f'{self.issn_electronic} {self.issn_print}'
+
+    class Meta:
+        unique_together = [
```
@robertatakenaka Comentário: parâmetro/variável interessante.

rafaelpezzuto commented on pid_provider/models.py:
```diff
@@ -0,0 +1,161 @@
+
+class XMLFile(models.Model):
+    uri = models.URLField(_("URI"), max_length=265, null=True, blank=False)
+    content = models.CharField('Content', max_length=64, null=True, blank=False)
```
@robertatakenaka Dúvida: 64 caracteres são o suficiente para esse campo?

rafaelpezzuto commented on pid_provider/models.py:
```diff
@@ -0,0 +1,161 @@
+    # return datetime.utcnow().isoformat().replace("T", " ") + "Z"
+
+
+class XMLJournal(models.Model):
```
@robertatakenaka Comentário: `null=True` significa que no banco de dados o campo poderá ser NULL. `blank=False` significa que nos formulários da aplicação o campo deverá ser preenchido. Então, fazer `null=True, blank=False` serviria apenas para algumas situações muito específicas que não sei se são os casos dos campos que você definiu neste e em outros modelos. Vide https://stackoverflow.com/questions/8609192/what-is-the-difference-between-null-true-and-blank-true-in-django.

rafaelpezzuto commented on collection/controller.py:
```diff
@@ -40,7 +40,10 @@ def update_files_storage_configuration(
+    try:
+        return FilesStorageConfiguration.objects.get(name=name)
+    except FilesStorageConfiguration.DoesNotExist:
+        return FilesStorageConfiguration(name=name)
```
@robertatakenaka Dúvida: você inicializa um novo FileStorageConfiguration cujo campo name é povoado com name. Porém, isso teria alguma utilizadade já que os outros campos estarão vazios? Não seria melhor retorar uma exceção para indicar que é preciso ter um FileStorageConfiguration configurado?

rafaelpezzuto commented on libs/xml_sps_utils.py:
```diff
@@ -0,0 +1,357 @@
+    starttag = endtag.replace("/", "").replace(">", "").strip()
+
+    p = xml_content.find(starttag)
+    return xml_content[:p], xml_content[p:].strip()
```
@robertatakenaka Testei com um XML e obtive, para o primeiro elemento retornado, a linha `<?xml version="1.0"... JATS-journalpublishing1.dtd">\n'`. Então suspeito que precisa fazer um `.strip()` no primeiro elemento.

rafaelpezzuto reviewed (commented):
Por enquanto olhei o código-fonte e fiz algumas observações. No momento tento fazer funcionar as funções que você indicou.

rafaelpezzuto commented on pid_provider/controller.py:
```diff
@@ -613,3 +617,47 @@ def _register_new_document(xml_adapter, xml_file, user_id):
+            # tentar recuperar documento do site novo
+            data = package_utils.get_article_data_for_comparison(xmltree)
+            similar_docs = publication_documents.get_similar_documents(
+                article_title=data["title"],
```
@robertatakenaka Nota de reunião: o title pode não ser bom para recuperar docs (por conta de caracteres não tratados)

rafaelpezzuto commented on pid_provider/controller.py:
```diff
@@ -0,0 +1,646 @@
+LOGGER_FMT = "%(asctime)s [%(levelname)s] %(name)s: %(message)s"
+
+
+def get_registered_xml(xml, user_id, fput_content):
```
Possibilitar `fput_content=None`

rafaelpezzuto commented on pid_provider/controller.py:
```diff
@@ -0,0 +1,646 @@
+        logging.info("ADAPT")
+        logging.info(xml_adapter)
+        registered = _query_document(xml_adapter)
+        if not registered:
```
`and fput_content`

rafaelpezzuto commented on pid_provider/controller.py:
```diff
@@ -0,0 +1,646 @@
+        registered = _query_document(xml_adapter)
+        if not registered:
+            _sync_new_website_to_pid_provider_system(
+                xml_adapter.xmltree, user_id, fput_content)
```
Criar task que itera nos docs do opac.article e gera dados internos no pid-provider.

rafaelpezzuto commented on pid_provider/controller.py:
```diff
@@ -0,0 +1,646 @@
+        )
+
+
+def request_document_ids_for_zip(xml_zip_file_path, user_id, fput_content):
```
Apagar

gitnnolabs commented:
@robertatakenaka 

Consegui baixar esse branch, criar as migrações e aplicar também, porém parei no ponto onde precisa criar um "FilesStorageConfiguration", acho que preciso de mais contexto para seguir com a validação!

gitnnolabs commented:
@robertatakenaka 

Estou tentando testar a função ``request_document_ids_for_zip``, porém na assinatura da função existe um parâmetro que não entendo o que devo passar ``fput_content``. 

```python
request_document_ids_for_zip(xml_zip_file_path, user_id, fput_content)
```

Se entendi é necessária uma instância do mínio da classe localizada em ``libs/dsm/files_storage/minio.py`` e seria injetar um método dessa classe correto? 

Produzi o seguinte script, por hora, para testar uma das funções: 

```python
# Cria uma instância do minio
m = minio.MinioStorage("192.168.15.78:9099", "PaVe1EZ1w2t8vGYd",
                        "cXLo4XrrWSIoN79wWnYavGTSsxLiOZfz", "root", "publication", False)

print([i for i in request_document_ids_for_zip("0001-3765-aabc-94-04-e20201438.xml", 1, m.fput_content)])
```  

Como resultado obtive o seguinte erro: 

![Captura de Tela 2022-11-07 às 07 39 29](https://user-images.githubusercontent.com/86991526/200290337-c8dcae78-101e-40a4-b173-0243e1cd047d.png)

Na função ``get_xml_items``, esta com referência para itens que não existem: 

![Captura de Tela 2022-11-07 às 07 40 28](https://user-images.githubusercontent.com/86991526/200290780-0b6c31ed-a617-4c54-9fa0-ae8bf5352d2c.png)

Bom ai tentei rodar com um arquivo .zip, já que nesse caso temos a função para .zip, alterei o script para: 

```python 

    # Cria uma instância do minio
    m = minio.MinioStorage("192.168.15.78:9099", "PaVe1EZ1w2t8vGYd",
                            "cXLo4XrrWSIoN79wWnYavGTSsxLiOZfz", "root", "publication", False)

    print([i for i in request_document_ids_for_zip("0001-3765-aabc-94-04-e20201438.zip", 1, m.fput_content)])

```

Nesse ponto percebi que temos erro de implementação o ``request_document_ids_for_zip`` sempre uma lista vazia.

gitnnolabs commented:
Testando a função ``get_registered_xml``: 

```python 

    # Cria uma instância do minio
    m = minio.MinioStorage("192.168.15.78:9099", "PaVe1EZ1w2t8vGYd",
                            "cXLo4XrrWSIoN79wWnYavGTSsxLiOZfz", "root", "publication", False)

    # Teste da função: request_document_ids_for_zip
    print([i for i in controller.request_document_ids_for_zip("0001-3765-aabc-94-04-e20201438.zip", 1, m.fput_content)])

    # Teste da função: get_registered_xml
    for pkg in models.Package.objects.all():
        for xml in xml_sps_utils.get_xml_items_from_zip_file(pkg.file.path):
            controller.get_registered_xml(xml, 1, m.fput_content)
```

Obtenho o seguinte erro: 

![Captura de Tela 2022-11-07 às 09 19 22](https://user-images.githubusercontent.com/86991526/200308768-1a16ff59-d6b4-42c2-bde4-4ebdb935a08b.png)

Vou corrigir para dá procedimento.

robertatakenaka committed:
Faz melhorias no nomes de classes e funções

robertatakenaka committed:
Remove a sincronização com os dados do site novo. Considerar que isso tenha sido feito previamente

robertatakenaka committed:
Torna as funções em atributos da classe PidProvider que são dependentes de files_storage

robertatakenaka committed:
Transfere algumas funções do módulo controller para atributos de BaseArticle
Cria módulo para registrar as requisições de pid

robertatakenaka committed:
Melhoria e correção no módulo genérico de XML SPS

- adiciona mais informação na mensagem de erro da exceção GetXmlWithPreError
- corrige o valor default para os atributos de artigo em fascículo

robertatakenaka committed:
Corrige problemas que impactam na identificação se o documento já está ou não registrado no Pid Provider

- remove códigos desnecessários
- corrige defeitos devido às mudanças de nomes de variáveis
- remove a conversão de valores None para _str_with_64_char

robertatakenaka committed:
Corrige a importação: troca funções de pid_provider/controller pela classe PidProvider

robertatakenaka committed:
Modifica os parâmetros para buscar a configuração do minio

robertatakenaka committed:
Faz correções no módulo pid_provider/controller

- corrige nomes de variáveis
- instancia MinioStorage ao instancia PidProvider
- corrige identificação de documentos registrados no pid provider e geração de valor para v2
- adiciona o atributo `request_document_ids_for_xml_uri` na classe PidProvider

robertatakenaka committed:
Cria script e tarefa para fazer a inserção dos documentos do site novo para o pid provider

robertatakenaka committed:
Cria uma app para fazer interface com minio

robertatakenaka committed:
Refatora o código para passar a usar a app files_storage

robertatakenaka committed:
Refatora o código do pid_provider para usar a app files_storage

robertatakenaka committed:
A função que dá a ordem do item no menu retorna um valor padrão quando não encontra a ordem definida

robertatakenaka committed:
Refatora a classe MigrationConfigurationController para melhorar o código e usar a app files_storage

robertatakenaka committed:
Considera que o retorno de migrate_issue_files pode retornar erro e, por isso, o status não deve ser MS_IMPORTED

robertatakenaka committed:
Inicia remodelagem para adequar o uso de MinioFile

robertatakenaka committed:
Remove o atributo migrate_issue_files de MigrationConfigurationController
Cria a classe IssueFilesController com atributos para migrar, adicionar, obter arquivos de issue files

robertatakenaka committed:
Transfere funções de get, create, update de JournalMigration, IssueMigration, DocumentMigration, de controller para atributos dos modelos

robertatakenaka committed:
Faz com que funções em controller passem a ser atributos dos modelos de journal

robertatakenaka committed:
Faz com que funções em controller passem a ser atributos dos modelos de issue

robertatakenaka committed:
Faz com que funções em controller sejam atributos nos modelos de collection

robertatakenaka committed:
Modifica as chamadas das funções de journal, issue e collection devido às mudanças nas funções de controller

robertatakenaka committed:
Transforma bloco dentro de `for` em uma função para fazer a migração de um documento

robertatakenaka committed:
Acrescenta atributos aos modelos da app collection

robertatakenaka committed:
Ajusta o módulo migration/controller.py para usar os atributos criados nos modelos da app collection

robertatakenaka committed:
Faz melhorias na classe FilesStorageManager, adicionando versões de `MinioFile`

robertatakenaka committed:
Cria a classe PidRequester para fazer o registro do XML no pid provider remoto e no local

robertatakenaka committed:
Cria a app publication para registrar os artigos publicados / publicáveis

robertatakenaka committed:
Cria a instância de PublicationArticle com os dados de um documento migrado

robertatakenaka committed:
Faz correções
